src/main/scala/edu/ie3/simona/model/em/opt/OptimizedFlexStrat.scala

```diff
@@ -0,0 +1,445 @@
+    * @param model
+    *   The optimization model to add variables and constraints to.
+    * @return
+    *   Containers that holds all results, including state and operation
+    *   variables.
+    */
+  def addAssetConstraints(
+      flexOptions: Iterable[(UUID, EnergyBoundariesFlexOptions)],
+      sampleTime: Time,
+      ticks: Seq[Long],
+  )(using model: MPModel): Iterable[AssetVarContainer] = {
+    flexOptions.map { case (assetUUID, fo) =>
+      val results = fo.energyBoundaries
+        .map(adaptEnergyBoundaries)
+        .map { boundaries =>
+          ticks.tail.foldLeft[IndexedSeq[StepResults]](IndexedSeq.empty) {
+            case (previousResults, tick) =>
+              val previousState = previousResults.lastOption.flatMap(_.state)
+
+              val res = addAssetStep(
+                boundaries,
+                tick,
+                sampleTime,
+                previousState,
+              )
+
+              previousResults.appended(res)
+          }
+        }
+
+      AssetVarContainer(assetUUID, results)
+    }
+  }
+
+  /** Creates and adds constraints for a single asset and time step. Relevant
+    * variables and soft constraints are returned in a [[StepResults]]
+    * container.
+    *
+    * @param energyBoundaries
+    *   The asset energy boundaries to use.
+    * @param tick
+    *   The tick to add constraints for.
+    * @param sampleTime
+    *   The amount of time between the steps.
+    * @param maybePreviousState
+    *   Optionally, the previous state variable.
+    * @param model
+    *   The optimization model to add variables and constraints to.
+    * @return
+    *   The results for this asset and time step.
+    */
+  private def addAssetStep(
+      energyBoundaries: AssetEnergyBoundaries,
+      tick: Long,
+      sampleTime: Time,
+      maybePreviousState: Option[Expression],
+  )(using model: MPModel): StepResults = {
+
+    val energyLimits = energyBoundaries.energyLimits
+      .maxBefore(tick + 1)
+      .map { case (_, limits) =>
+        limits
+      }
+      .getOrElse(throw new CriticalFailureException("No energy limits found!"))
+
+    val formerEnergyLimits =
+      energyBoundaries.energyLimits.maxBefore(tick).map { case (_, limits) =>
+        limits
+      }
+
+    if energyLimits.getLower == energyLimits.getUpper &&
+      formerEnergyLimits.forall(limits => limits.getLower == limits.getUpper)
+    then {
+      // there is no flexibility at all, thus we don't need any state to keep track of
+
+      val formerEnergy = formerEnergyLimits.map(_.getUpper).getOrElse(zeroKWh)
+      val currentEnergy = energyLimits.getUpper
+
+      val fixedPower = (currentEnergy - formerEnergy) / sampleTime
+      StepResults(Const(fixedPower.toKilowatts), None, None)
+    } else {
+      // we do have some flexibility at this point in time, model it
+
+      // we use charging efficiency for both charging and discharging,
+      // since we use the adapted storage model here
```

Comment:
Perhaps you could find a better description for 'the adapted storage model'. I'm not sure if we'll be able to match this naming in the future.

Reply:
I could've called it "linearized storage model", but since I don't have a monopoly on linearizing storage models, I thought this would not help much either. I also thought it might be clear what adaptation is meant here, given that I explain this in the scaladoc of `adaptEnergyBoundaries`.